include/picongpu/particles/externalBeam/density/ProbingBeamImpl.hpp

```diff
@@ -0,0 +1,205 @@
+        static constexpr float_X reducedCellCenter_x{reducedCellCenterBeam[Side::template SimToBeamIdx_t<0u>::value]};
+        static constexpr float_X reducedCellCenter_y{reducedCellCenterBeam[Side::template SimToBeamIdx_t<1u>::value]};
+        static constexpr float_X reducedCellCenter_z{reducedCellCenterBeam[Side::template SimToBeamIdx_t<2u>::value]};
+
+
+        // the area of the cell side perpendicular to the propagation
+        static constexpr float_64 CELL_AREA_SI = detail::GetCellAreaSI<Side>::get();
+        static constexpr float_X PHOTONS_IN_A_CELL
+            = static_cast<float_X>(photonFluxAtMaxBeamIntensity * SI::DELTA_T_SI * CELL_AREA_SI);
+        // the photon density in the cell in terms of the base density
+        static constexpr float_X REFERENCE_PHOTON_DENSITY = PHOTONS_IN_A_CELL / CELL_VOLUME / BASE_DENSITY;
+
+    public:
+        template<typename T_SpeciesType>
+        struct apply
+        {
+            using type = ProbingBeamImpl<ParamClass>;
+        };
+        /* Host side constructor
+         *
+         * Provides domain info used for checking if the functor is at a simulation boundary
+         */
+        HINLINE ProbingBeamImpl(uint32_t const& currentStep) : probingBeam_m(), currentStep_m(currentStep)
+        {
+            // TODO: modernize it offset  should be in subGrid already?
+            uint32_t const numSlides = MovingWindow::getInstance().getSlideCounter(currentStep);
+            SubGrid<simDim> const& subGrid = Environment<simDim>::get().SubGrid();
+            DataSpace<simDim> const localCells = subGrid.getLocalDomain().size;
+            globalDomain_m = subGrid.getGlobalDomain();
+            globalDomain_m.offset.y() += numSlides * localCells.y();
+        }
+
+        //! Check if we are in the first row of cells and at the side where the particles are injected
+        DINLINE bool isInjectionBoundary(const DataSpace<simDim>& totalCellOffset)
+        {
+            const auto globalCellOffset = totalCellOffset - globalDomain_m.offset;
+            // Propagation axis
+            constexpr auto d{Side::template BeamToSimIdx_t<2u>::value};
+            bool boundary = false;
+            if constexpr(reverse)
+            {
+                boundary = globalCellOffset[d] == globalDomain_m.size[d] - 1;
+            }
+            else
+            {
+                boundary = globalCellOffset[d] == 0;
+            }
+            return boundary;
+        }
+
+        /** Calculate the normalized density
+         *
+         * @param totalCellOffset total offset including all slides [in cells]
+         */
+        DINLINE float_X operator()(const DataSpace<simDim>& totalCellOffset)
+        {
+            // Do not create any particles beside the injection cells
+            if(not isInjectionBoundary(totalCellOffset))
+            {
+                return 0.0_X;
+            }
+
+            float3_X totalCellOffsetVector;
+            if constexpr(simDim == 2u)
```

Comment:
ooks like a compiler bug is failing to see that if `simDim` is `2u` the else branch will not be executed.
suggested

```C++
totalCellOffsetVector.x() = totalCellOffset.x();
totalCellOffsetVector.y() = totalCellOffset.y();
if constexpr(simDim == 2)
    totalCellOffsetVector[2] = 0.0_X;
else
    totalCellOffsetVector[2] = totalCellOffset[2];
```